### PR TITLE
Promote IO support queries to cudf API

### PR DESCRIPTION
**Pull request by robertmaynard**
## Description
Promote the ability to query the status of cufile and nvcomp support to the public API. It seems like these kind of questions would want to be asked by external users. 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/io/nvcomp_adapter.hpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+#include <optional>
+#include <string>
+
+namespace CUDF_EXPORT cudf {
+namespace io::nvcomp {
+
+enum class compression_type { SNAPPY, ZSTD, DEFLATE, LZ4 };
+
+/**
+ * @brief Set of parameters that impact whether the use nvCOMP features is enabled.
```
Comment (bdice) [inline cpp/include/cudf/io/nvcomp_adapter.hpp:30]:
```suggestion
 * @brief Set of parameters that impact whether nvCOMP features are enabled.
```

File: cpp/include/cudf/io/nvcomp_adapter.hpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  bool are_all_integrations_enabled;     ///< all integrations
+  bool are_stable_integrations_enabled;  ///< stable integrations
+  int compute_capability_major;          ///< cuda compute major version
+
+  /**
+   * @brief Default Constructor
+   */
+  feature_status_parameters();
+
+  /**pht
```
Comment (bdice) [inline cpp/include/cudf/io/nvcomp_adapter.hpp:46]: Typo? Not sure what this means.
```suggestion
  /**
```

File: cpp/include/cudf/io/nvcomp_adapter.hpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+   * @brief feature_status_parameters Constructor
+   *
+   * @param major positive integer representing major value
+   * @param minor positive integer representing minor value
+   * @param patch positive integer representing patch value
```
Comment (bdice) [inline cpp/include/cudf/io/nvcomp_adapter.hpp:51]:
```suggestion
   * @param major positive integer representing major value of nvcomp
   * @param minor positive integer representing minor value of nvcomp
   * @param patch positive integer representing patch value of nvcomp
```

File: cpp/include/cudf/io/nvcomp_adapter.hpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+   * @param all_enabled if all integrations are enabled
+   * @param stable_enabled if stable integrations are enabled
+   * @param cc_major cuda compute capability
```
Comment (bdice) [inline cpp/include/cudf/io/nvcomp_adapter.hpp:54]:
```suggestion
   * @param cc_major CUDA compute capability
```

File: cpp/include/cudf/io/nvcomp_adapter.hpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+         lhs.lib_patch_version == rhs.lib_patch_version and
+         lhs.are_all_integrations_enabled == rhs.are_all_integrations_enabled and
+         lhs.are_stable_integrations_enabled == rhs.are_stable_integrations_enabled and
+         lhs.compute_capability_major == rhs.compute_capability_major;
+}
+
+/**
+ * @brief If a compression type is disabled through nvCOMP, returns the reason as a string.
+ *
+ * Result cab depend on nvCOMP version and environment variables.
```
Comment (bdice) [inline cpp/include/cudf/io/nvcomp_adapter.hpp:84]:
```suggestion
 * Result depends on nvCOMP version and environment variables.
```

File: cpp/include/cudf/io/nvcomp_adapter.hpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+ * @param params Optional parameters to query status with different configurations
+ * @returns Reason for the feature disablement, `std::nullopt` if the feature is enabled
+ */
+[[nodiscard]] std::optional<std::string> is_compression_disabled(
+  compression_type compression, feature_status_parameters params = feature_status_parameters());
+
+/**
+ * @brief If a decompression type is disabled through nvCOMP, returns the reason as a string.
+ *
+ * Result can depend on nvCOMP version and environment variables.
```
Comment (bdice) [inline cpp/include/cudf/io/nvcomp_adapter.hpp:96]:
```suggestion
 * Result depends on nvCOMP version and environment variables.
```

**Review by bdice — APPROVED** [review]
Just a few doc nits, otherwise LGTM.

**Review by mhaseeb123 — APPROVED** [review]

**Pull request merged**